trieste/models/gpflux/architectures.py

```diff
@@ -0,0 +1,175 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""
+This module provides :func:`build_constant_input_dim_deep_gp` to build a Deep GP of
+arbitrary depth where each hidden layer has the same input dimensionality as the data.
+"""
+
+from dataclasses import dataclass
+from typing import cast
+
+import numpy as np
+import tensorflow as tf
+from scipy.cluster.vq import kmeans2
+
+import gpflow
+from gpflow.kernels import SquaredExponential
+from gpflow.likelihoods import Gaussian
+
+from gpflux.helpers import (
+    construct_basic_inducing_variables,
+    construct_basic_kernel,
+    construct_mean_function,
+)
+from gpflux.layers.gp_layer import GPLayer
+from gpflux.layers.likelihood_layer import LikelihoodLayer
+from gpflux.models import DeepGP
+
+
+@dataclass
+class Config:
+    """
+    The configuration used by :func:`build_constant_input_dim_deep_gp`.
+    """
+
+    num_inducing: int
+    """
+    The number of inducing variables, *M*. The Deep GP uses the same number
+    of inducing variables in each layer.
+    """
+
+    inner_layer_qsqrt_factor: float
+    """
+    A multiplicative factor used to rescale the hidden layers'
+    :attr:`~gpflux.layers.GPLayer.q_sqrt`. Typically this value is chosen to be small
+    (e.g., 1e-5) to reduce noise at the start of training.
+    """
+
+    likelihood_noise_variance: float
+    """
+    The variance of the :class:`~gpflow.likelihoods.Gaussian` likelihood that is used
+    by the Deep GP.
+    """
+
+    whiten: bool = True
+    """
+    Determines the parameterisation of the inducing variables.
+    If `True`, :math:``p(u) = N(0, I)``, otherwise :math:``p(u) = N(0, K_{uu})``.
+    .. seealso:: :attr:`gpflux.layers.GPLayer.whiten`
+    """
+
+
+def _construct_kernel(input_dim: int, is_last_layer: bool) -> SquaredExponential:
+    """
+    Return a :class:`gpflow.kernels.SquaredExponential` kernel with ARD lengthscales set to
+    2 and a small kernel variance of 1e-6 if the kernel is part of a hidden layer;
+    otherwise, the kernel variance is set to 1.0.
+
+    :param input_dim: The input dimensionality of the layer.
+    :param is_last_layer: Whether the kernel is part of the last layer in the Deep GP.
+    """
+    variance = 1e-6 if not is_last_layer else 1.0
+
+    # TODO: Looking at this initializing to 2 (assuming N(0, 1) or U[0,1] normalized
+    # data) seems a bit weird - that's really long lengthscales? And I remember seeing
+    # something where the value scaled with the number of dimensions before
+    lengthscales = [2.0] * input_dim
+    return SquaredExponential(lengthscales=lengthscales, variance=variance)
+
+def build_constant_input_dim_flexible_deep_gp(X: np.ndarray, num_layers: int, config: Config, separate_case: bool, output_dim: bool) -> DeepGP:
```

Comment:
this seems like it should go to gpflux rather than being here in trieste, if its generally useful and not just for trieste needs

Reply:
I will probably move this architecture to tests..since that's the only place it will be ever used

Reply:
great!